bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3ChunkStorage.java

```diff
@@ -0,0 +1,440 @@
+        return 0;
+    }
+
+    /**
+     * Concatenates two or more chunks using storage native  functionality. (Eg. Multipart upload.)
+     *
+     * @param chunks Array of ConcatArgument objects containing info about existing chunks to be appended together.
+     *               The chunks are appended in the same sequence the names are provided.
+     * @return int Number of bytes concatenated.
+     * @throws ChunkStorageException         Throws ChunkStorageException in case of I/O related exceptions.
+     * @throws UnsupportedOperationException If this operation is not supported by this provider.
+     */
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException, UnsupportedOperationException {
+        int totalBytesConcated = 0;
+        try {
+            int partNumber = 1;
+
+            SortedSet<MultipartPartETag> partEtags = new TreeSet<>();
+            String targetPath = config.getPrefix() + chunks[0].getName();
+            String uploadId = client.initiateMultipartUpload(config.getBucket(), targetPath);
+
+            // check whether the target exists
+            if (!checkExists(chunks[0].getName())) {
+                throw new FileNotFoundException(chunks[0].getName());
+            }
+
+            //Copy the parts
+            for (int i = 0; i < chunks.length; i++) {
+                if (0 != chunks[i].getLength()) {
+                    val sourceHandle = chunks[i];
+                    S3ObjectMetadata metadataResult = client.getObjectMetadata(config.getBucket(),
+                            config.getPrefix() + sourceHandle.getName());
+                    long objectSize = metadataResult.getContentLength(); // in bytes
+                    Preconditions.checkState(objectSize >= chunks[i].getLength());
+                    CopyPartRequest copyRequest = new CopyPartRequest(config.getBucket(),
+                            config.getPrefix() + sourceHandle.getName(),
+                            config.getBucket(),
+                            targetPath,
+                            uploadId,
+                            partNumber++).withSourceRange(Range.fromOffsetLength(0, chunks[i].getLength()));
+
+                    CopyPartResult copyResult = client.copyPart(copyRequest);
+                    partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+                    totalBytesConcated += chunks[i].getLength();
+                }
+            }
+
+            //Close the upload
+            client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
+                    targetPath, uploadId).withParts(partEtags));
+            // Delete all source objects.
+            for (int i = 1; i < chunks.length; i++) {
+                client.deleteObject(config.getBucket(), config.getPrefix() + chunks[i].getName());
+            }
+        } catch (RuntimeException e) {
+            throw e; // make spotbugs happy
```

Comment:
Is spotbugs really complaining about this? We don't do it in other parts of the code and it doesn't complain, why does it complain here, do you know?

Reply:
I'm not sure why